### PR TITLE
templates: use pg_isready as liveness probe

### PR DESCRIPTION
**Pull request by pkubatrh**
Resolves: #195

### CODE REVIEW

**Comment by pkubatrh** [comment]
@praiskup Can you take a quick look at this?

@pvalena @bparees Is there anything more we need to do to get the change into the Openshift templates?

**Comment by pvalena** [comment]
AFAIK you primarily need to add the new version `9.6`

- as a default for deployment; and
- in all texts, such as ` 9.2, 9.4, 9.5 or latest `

Thanks!

**Comment by pkubatrh** [comment]
Makes sense, thanks.

File: examples/postgresql-ephemeral-template.json [hunk 1]
```diff
@@ -245,8 +245,8 @@
     {
       "name": "POSTGRESQL_VERSION",
       "displayName": "Version of PostgreSQL Image",
-      "description": "Version of PostgreSQL image to be used (9.2, 9.4, 9.5 or latest).",
-      "value": "9.5",
+      "description": "Version of PostgreSQL image to be used (9.4, 9.5, 9.6 or latest).",
```
Comment (pvalena) [inline examples/postgresql-ephemeral-template.json]: Is 9.2 not relevant anymore?

Comment (pkubatrh) [reply examples/postgresql-ephemeral-template.json]: Has been deprecated for a long while already.

File: examples/postgresql-ephemeral-template.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
     "name": "postgresql-ephemeral",
     "annotations": {
       "openshift.io/display-name": "PostgreSQL (Ephemeral)",
-      "description": "PostgreSQL database service, without persistent storage. For more information about using this template, including OpenShift considerations, see https://github.com/sclorg/postgresql-container/blob/master/9.5.\n\nWARNING: Any data stored will be lost upon pod destruction. Only use this template for testing",
+      "description": "PostgreSQL database service, without persistent storage. For more information about using this template, including OpenShift considerations, see https://github.com/sclorg/postgresql-container/.\n\nWARNING: Any data stored will be lost upon pod destruction. Only use this template for testing",
```
Comment (pvalena) [inline examples/postgresql-ephemeral-template.json]: AFAIK this was meant to give container-specific instructions (real simplistic), thus reading `README.md` - usage - for direct usage of OpenShift; regardless of version(most ppl just use `latest`). IOW: I'd rather keep the 9.6.

Comment (pvalena) [reply examples/postgresql-ephemeral-template.json]: f.e. https://github.com/sclorg/postgresql-container/tree/master/9.5#environment-variables-and-volumes

File: examples/postgresql-ephemeral-template.json (hunk 1 above)
Comment (pvalena) [reply examples/postgresql-ephemeral-template.json]: Ok. We'll need to keep an eye for that.

File: examples/postgresql-ephemeral-template.json (hunk 2 above)
Comment (pkubatrh) [reply examples/postgresql-ephemeral-template.json]: I have modified this in preparation for the last stages of image content generation (see #203 ) - eventually there will not be a `9.6` to directly link to and so the current link would result in a 404.

Comment (pkubatrh) [reply examples/postgresql-ephemeral-template.json]: Additionally, it is currently impossible to link to the 9.6 README anyway, without going through `latest/` which might change into a different version in the future. So it makes sense to me to just point users to the root.

**Comment by pvalena** [comment]
-- I hope it's not a problem for OpenShift, as they might link to it somewhere AFAIK.

@bparees, is it fine if `9.6` folder is just a symlink? Can we link root, or possibly `latest` directory?

**Comment by pvalena** [comment]
FTR: Subsequently I'll create image-streams and PR for openshift/origin.

**Comment by pvalena** [comment]
Also: *before* this PR gets merged I'd like to have the original template tested it by CI in OpenShift, so the changes can be actually validated.

See https://github.com/sclorg/postgresql-container/issues/204.

File: examples/postgresql-ephemeral-template.json (hunk 1 above)
Comment (praiskup) [reply examples/postgresql-ephemeral-template.json]: pg_isready is not available for 9.2 anyway

File: examples/postgresql-ephemeral-template.json [hunk 3]
```diff
@@ -127,11 +127,11 @@
                   }
                 },
                 "livenessProbe": {
+                  "exec": {
+                    "command": [ "/bin/sh", "-i", "-c", "pg_isready -h 127.0.0.1 -p 5432" ]
```
Comment (praiskup) [inline examples/postgresql-ephemeral-template.json]: I guess the `/bin/sh` is necessary, but I would rather avoid using the `-i` option, if possible.

Comment (pkubatrh) [reply examples/postgresql-ephemeral-template.json]: I have left it in only because it was present in other /bin/sh invocations inside the template

**Comment by praiskup** [comment]
> @bparees, is it fine if 9.6 folder is just a symlink? Can we link root, or possibly latest directory?

This is long time discussed thing;  IMO linking directly to gitrepo - to concrete paths - is a bad practice.  There's a card on trello where we could discuss/implement a better place for documentation.

Regarding the link, it is a temporary issue ... since we could move the "latest" to PostgreSQL 10 soon, at least for testing purposes.

**Comment by praiskup** [comment]
This looks good to me, but having this tested is a must :-)

File: examples/replica/postgresql_replica.json [hunk 4]
```diff
@@ -71,7 +71,7 @@
       "name": "IMAGESTREAMTAG",
       "displayName": "ImageStreamTag",
       "description": "The OpenShift ImageStreamTag to use for PostgreSQL.",
-      "value": "postgresql:9.5"
+      "value": "postgresql:9.6"
```
Comment (bparees) [inline examples/replica/postgresql_replica.json]: there's no postgres 9.6 imagestream tag in openshift yet:
https://github.com/openshift/origin/blob/master/examples/image-streams/image-streams-centos7.json#L724-L795

so you need to deliver that before you start making templates default to looking for it.

**Comment by bparees** [comment]
@praiskup can you lay out the plan/order of operations here?  is it to:

1) deliver this here
2) deliver the 9.6 imagestream to origin/image-streams
3) deliver this template to origin/db-templates

?

File: examples/replica/postgresql_replica.json (hunk 4 above)
Comment (pvalena) [reply examples/replica/postgresql_replica.json]: We'll update at once - in our repos. The template will be actually tested with freshly build and tagged image into custom namespace.

**Comment by praiskup** [comment]
According to the discussion here, I think so :-), per @pvalena comment:

>  FTR: Subsequently I'll create image-streams and PR for openshift/origin.

**Comment by pvalena** [comment]
@bparees - The order is a bit parallelized, but basically:

 1. openshift tests
 2. updated template (container folder already done)
 3. generated [image-streams](https://github.com/sclorg/image-streams/tree/staging), although currently using [manual comparison](https://github.com/sclorg/image-streams/tree/origin)
 4. Pull Requests to openshift/origin

Where (1+2) and 3 are being worked on in parallel (I'm working on the image-streams; and together with @hhorak will provide help with the steps above).

**Comment by pkubatrh** [comment]
I will look into getting the tests up and running so we can test this out properly.

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
@praiskup Seems like resalloc does not want to cooperate again.

**Comment by praiskup** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
Thanks! Tests are green now

**Comment by pkubatrh** [comment]
Fixed one more issue with pure image tests [test-openshift]

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by praiskup** [comment]
lgtm

**Comment by pkubatrh** [comment]
Thanks for review! I will make one more rebase against container common scripts to pick up https://github.com/sclorg/container-common-scripts/pull/45

It is not needed for postgresql's openshift tests right now but I would like to have it fixed for eventual s2i (or other cases accessing the internet) before the problem is forgotten again.

**Comment by pkubatrh** [comment]
[test-openshift]

**Comment by pkubatrh** [comment]
Tests are green. Merging this in. Thanks everyone!

**Pull request merged**